Add Karn Keyboard

robcmills opened the pull request:
## Description

Added a new folder `keyboards/karn`, with the necessary files to build the firmware for the Karn keyboard.

https://github.com/robcmills/karn-keyboard

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

robcmills commented:
This keyboard and PR is based on a very similar keyboard (the Cantor) that has already been added: https://github.com/qmk/qmk_firmware/pull/16552

waffle87 commented on keyboards/karn/config.h:
```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include "config_common.h"
+
+/* key matrix size */
+#define MATRIX_ROWS 8
+#define MATRIX_COLS 6
```
```suggestion
```

waffle87 commented on keyboards/karn/rules.mk:
```diff
@@ -0,0 +1,3 @@
+SPLIT_KEYBOARD = yes
+SERIAL_DRIVER = usart
+TAP_DANCE_ENABLE = yes
```
This can't live at the keyboard level, as it will cause compilation errors when using the QMK Configurator or any other keymap that doesn't have the necessary tap dance code.

robcmills replied:
@waffle87 Ok thanks for letting me know. Where should I move it to? (trying to figure this out from the docs but having a hard time)

robcmills replied:
Trying moving just the `TAP_DANCE_ENABLE` to `keyboards/karn/keymaps/colemak/rules.mk`

waffle87 commented on keyboards/karn/readme.md:
```diff
@@ -0,0 +1,99 @@
+
+![karn-2-full](https://github.com/robcmills/karn-keyboard/blob/main/images/karn-2-full.jpg)
+
+![karn-2-pcb](https://github.com/robcmills/karn-keyboard/blob/main/images/karn-2-pcb.png)
```
These images should be rehosted on Imgur.

waffle87 commented on keyboards/karn/readme.md:
```diff
@@ -0,0 +1,99 @@
+
+[https://github.com/robcmills/karn-keyboard](https://github.com/robcmills/karn-keyboard)
+
+![karn-2-left](https://github.com/robcmills/karn-keyboard/blob/main/images/karn-2-left.jpg)
+
+![karn-2-full](https://github.com/robcmills/karn-keyboard/blob/main/images/karn-2-full.jpg)
+
+![karn-2-pcb](https://github.com/robcmills/karn-keyboard/blob/main/images/karn-2-pcb.png)
```
These images should be rehosted on Imgur.

robcmills replied:
[karn 90dcbd1ff6] Host images on imgur

waffle87 commented on keyboards/karn/keymaps/colemak/keymap.c:
```diff
@@ -0,0 +1,99 @@
+    [TD_F2_F12] = ACTION_TAP_DANCE_DOUBLE(KC_F2, KC_F12),
+};
+
+// Equivalent to zmk behavior-hold-tap hold-preferred flavor
+#define HOLD_ON_OTHER_KEY_PRESS_PER_KEY
+
```
```suggestion
```

robcmills replied:
[karn 1b888a811b] Remove #define left in colemak/keymap.c

zvecr commented on keyboards/karn/info.json:
```diff
@@ -0,0 +1,71 @@
+    "manufacturer": "Robert Mills",
+    "keyboard_name": "karn",
+    "maintainer": "robcmills",
+    "bootloader": "stm32-dfu",
```
```suggestion
```

robcmills replied:
@zvecr Removing this causes the build to fail: https://github.com/qmk/qmk_firmware/runs/11340797021
Where should the bootloader be specified?

robcmills replied:
Welp, looks like the subsequent commit to change the dev board fixed it? https://github.com/qmk/qmk_firmware/pull/19846/commits/a37bb6ff00516027e5d03303b66e730785b325dc
Anyways, looks like builds are passing now 👍

zvecr commented on keyboards/karn/info.json:
```diff
@@ -0,0 +1,71 @@
+        "mousekey": true,
+        "nkro": true
+    },
+    "processor": "STM32F401",
+    "board": "BLACKPILL_STM32_F401",
```
```suggestion
    "development_board": "blackpill_f401",
```

zvecr commented on keyboards/karn/config.h:
```diff
@@ -0,0 +1,40 @@
+/* Top right key on right half */
+#define BOOTMAGIC_LITE_ROW_RIGHT    4
+#define BOOTMAGIC_LITE_COLUMN_RIGHT 4
+
+// Equivalent to zmk behavior-hold-tap tap-preferred flavor
+// Do not force the mod-tap key press to be handled as a modifier
+// if any other key was pressed while the mod-tap key is held down.
+#define IGNORE_MOD_TAP_INTERRUPT_PER_KEY
+
+// Equivalent to zmk behavior-hold-tap hold-preferred flavor
+#define HOLD_ON_OTHER_KEY_PRESS_PER_KEY
+
```
Shouldnt be enabled at the keyboard level, and should be moved to keymaps where required.
```suggestion
```

robcmills replied:
@zvecr Ok, not sure if this is correct, but moved these to `keyboards/karn/keymaps/colemak/config.h` (which is my customized keymap). And I cleaned up the default keymap to not include any of these defines or tap dance.

[karn 88d9db02bb]

zvecr commented on keyboards/karn/readme.md:
````diff
@@ -0,0 +1,99 @@
+  *       │  \  │  `  │  '  │  "  │  {  │     SYM     │  }  │  -  │  =  │  [  │  ]  │
+  *       │     │     │     │     │     │    LAYER    │     │     │     │     │     │
+  * ┌─────┼─────┼─────┼─────┼─────┼─────┤             ├─────┼─────┼─────┼─────┼─────┼─────┐
+  * │     │  !  │  @  │  #  │  $  │  %  │             │  ^  │  &  │  *  │  (  │  )  │     │
+  * │     │     │     │     │     │     │             │     │     │     │     │     │     │
+  * └─────┼─────┼─────┼─────┼─────┼─────┤             ├─────┼─────┼─────┼─────┼─────┼─────┘
+  *       │     │     │     │     │     │             │     │     │     │     │     │
+  *       │     │     │     │     │     │             │     │     │     │     │     │
+  *       └─────┴─────┴─────┼─────┼─────┼─────┐ ┌─────┼─────┼─────┼─────┴─────┴─────┘
+  *                         │     │     │     │ │ DEL │     │     │
+  *                         │     │     │     │ │     │     │     │
+  *                         └─────┴─────┴─────┘ └─────┴─────┴─────┘
+  *       ┌─────┬─────┬─────┬─────┬─────┐             ┌─────┬─────┬─────┬─────┬─────┐
+  *       │     │     │     │     │ VOL │     NAV     │     │     │     │     │     │
+  *       │     │     │     │     │ UP  │    LAYER    │     │     │     │     │     │
+  * ┌─────┼─────┼─────┼─────┼─────┼─────┤             ├─────┼─────┼─────┼─────┼─────┼─────┐
+  * │     │ CTL │ CTL │ CMD │ CMD │ VOL │             │  ←  │  ↓  │  ↑  │  →  │CAPS │     │
+  * │     │ + ↓ │ + ↑ │ + ` │ +TAB│ DOWN│             │     │     │     │     │LOCK │     │
+  * └─────┼─────┼─────┼─────┼─────┼─────┤             ├─────┼─────┼─────┼─────┼─────┼─────┘
+  *       │     │     │     │     │ MUTE│             │     │     │     │     │     │
+  *       │     │     │     │     │     │             │     │     │     │     │     │
+  *       └─────┴─────┴─────┼─────┼─────┼─────┐ ┌─────┼─────┼─────┼─────┴─────┴─────┘
+  *                         │     │     │     │ │     │     │     │
+  *                         │     │     │     │ │     │     │     │
+  *                         └─────┴─────┴─────┘ └─────┴─────┴─────┘
+  *       ┌─────┬─────┬─────┬─────┬─────┐             ┌─────┬─────┬─────┬─────┬─────┐
+  *       │ F1  │ F2  │ F3  │ F4  │ F5  │     NUM     │ F6  │ F7  │ F8  │ F9  │ F10 │
+  *       │ F11 │ F12 │     │     │     │    LAYER    │     │     │     │     │     │
+  * ┌─────┼─────┼─────┼─────┼─────┼─────┤   ┌─────┐   ├─────┼─────┼─────┼─────┼─────┼─────┐
+  * │     │  1  │  2  │  3  │  4  │  5  │   │ TAP │   │  6  │  7  │  8  │  9  │  0  │     │
+  * │     │     │     │     │     │     │   │TAPx2│   │     │     │     │     │     │     │
+  * └─────┼─────┼─────┼─────┼─────┼─────┤   └─────┘   ├─────┼─────┼─────┼─────┼─────┼─────┘
+  *       │     │     │     │     │     │             │     │     │     │     │     │
+  *       │     │     │     │     │     │             │     │     │     │     │     │
+  *       └─────┴─────┴─────┼─────┼─────┼─────┐ ┌─────┼─────┼─────┼─────┴─────┴─────┘
+  *                         │     │     │     │ │     │     │     │
+  *                         │     │     │     │ │     │     │     │
+  *                         └─────┴─────┴─────┘ └─────┴─────┴─────┘
+  */
+```
````
Shouldnt be here, and moved to a `keymaps/default/readme.md`

robcmills replied:
[karn 1afcf5e783] Move keymap ascii diagram from root readme to keymaps/colemak/readme

robcmills commented:
@zvecr Thank you for reviewing this. I think I have addressed all your review comments. I resolved the comment threads that seemed straightforward, and left the others open for your review. Please let me know if there are any other changes needed. 🙏

drashna commented on keyboards/karn/readme.md:
```diff
@@ -0,0 +1,42 @@
+
+[https://github.com/robcmills/karn-keyboard](https://github.com/robcmills/karn-keyboard)
+
+![karn-2-left](https://i.imgur.com/vm6XFyI.jpeg)
+
+![karn-2-full](https://i.imgur.com/R7WnlC3.jpeg)
+
+![karn-2-pcb](https://i.imgur.com/EQqyyED.png)
```
This uses a smaller, but high res version of the images, that is more suited to github's markdown
```suggestion
![karn-2-left](https://i.imgur.com/vm6XFyIh.jpeg)

![karn-2-full](https://i.imgur.com/R7WnlC3h.jpeg)

![karn-2-pcb](https://i.imgur.com/EQqyyEDh.png)
```

robcmills commented:
@zvecr @drashna Thanks for the reviews. I am experiencing an odd bug with Github whereby if I request review from one of you, it removes the review request from the other automatically. I can't seem to get it into a state where review is requested from both. 
Anyway, I think this PR is ready for another (final?) review.

drashna commented on keyboards/karn/keymaps/colemak/config.h:
```diff
@@ -0,0 +1,11 @@
+// Copyright 2023 Robert Mills (@robcmills)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+// Equivalent to zmk behavior-hold-tap tap-preferred flavor
```
```suggestion
#pragma once

// Equivalent to zmk behavior-hold-tap tap-preferred flavor
```

waffle87 commented on keyboards/karn/info.json:
```diff
@@ -0,0 +1,69 @@
+        "device_version": "1.0.0",
+        "pid": "0x0000",
+        "vid": "0xFEED"
+    },
```
```suggestion
    },
    "bootmagic": {
        "matrix": [0, 1]
    },
    "split": {
        "enabled": true,
        "bootmagic": {
            "matrix": [4, 4]
        }
    },
```

robcmills replied:
So it looks like there is a general move from the `config.h` to conveying everything in `info.json`. Is that an accurate description of the current state of qmk?

waffle87 commented on keyboards/karn/config.h:
```diff
@@ -0,0 +1,26 @@
+#define SERIAL_USART_FULL_DUPLEX   // Enable full duplex operation mode.
+#define SERIAL_USART_TX_PIN B6     // USART TX pin
+#define SERIAL_USART_RX_PIN B7     // USART RX pin
+
+/* Top left key on left half */
+#define BOOTMAGIC_LITE_ROW    0
+#define BOOTMAGIC_LITE_COLUMN 1
+/* Top right key on right half */
+#define BOOTMAGIC_LITE_ROW_RIGHT    4
+#define BOOTMAGIC_LITE_COLUMN_RIGHT 4
+
```
```suggestion
```

waffle87 commented on keyboards/karn/rules.mk:
```diff
@@ -0,0 +1,2 @@
+SPLIT_KEYBOARD = yes
```
```suggestion
```

robcmills commented:
@waffle87 Looks like some of the changes you suggested caused the build to fail:

https://github.com/qmk/qmk_firmware/pull/19846/checks?check_run_id=11439752925

```
Ψ Compiling keymap with gmake --jobs=41 --output-sync=target karn:default


QMK Firmware 0.19.12
☒ Not including data from file: keyboards/karn/info.json
☒ 	split: Additional properties are not allowed ('bootmagic' was unexpected)
☒ karn: No LAYOUTs defined! Need at least one layout defined in info.json.
Making karn with keymap default

☒ Not including data from file: keyboards/karn/info.json
☒ 	split: Additional properties are not allowed ('bootmagic' was unexpected)
☒ karn: No LAYOUTs defined! Need at least one layout defined in info.json.

 * No bootloader specified. Please set an appropriate bootloader in your rules.mk or info.json.        [ERRORS]

platforms/avr/bootloader.mk:142: *** Invalid BOOTLOADER.  Stop.
gmake: *** [Makefile:414: karn:default] Error 1
```

I will try to fix...

waffle87 commented:
Ah my bad — looks like `split.bootmagic` key only exists on `develop` currently.

robcmills commented:
@zvecr @drashna No rush, but still waiting on approval of this PR. Just pinging before it gets too stale. Thanks 🙏
